hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieMetadataTableValidator.java

```diff
@@ -0,0 +1,506 @@
+    } else {
+      LOG.info("Validation of getLatestBaseFiles success.");
+    }
+  }
+
+  /**
+   * Compare getLatestFileSlices between metadata table and fileSystem.
+   */
+  private void validateLatestFileSlices(HoodieTableFileSystemView metaFsView, HoodieTableFileSystemView fsView, String partitionPath) {
+
+    List<FileSlice> latestFileSlicesFromMetadataTable = metaFsView.getLatestFileSlices(partitionPath).sorted(new FileSliceCompactor()).collect(Collectors.toList());
+    List<FileSlice> latestFileSlicesFromFS = fsView.getLatestFileSlices(partitionPath).sorted(new FileSliceCompactor()).collect(Collectors.toList());
+
+    LOG.info("Latest file list from metadata: " + latestFileSlicesFromMetadataTable + ". For partition " + partitionPath);
+    LOG.info("Latest file list from direct listing: " + latestFileSlicesFromFS + ". For partition " + partitionPath);
+
+    validateFileSlice(latestFileSlicesFromMetadataTable, latestFileSlicesFromFS, partitionPath);
+    LOG.info("Validation of getLatestFileSlices success.");
+  }
+
+  private HoodieTableFileSystemView createHoodieTableFileSystemView(HoodieSparkEngineContext engineContext, boolean enableMetadataTable) {
+
+    HoodieMetadataConfig metadataConfig = HoodieMetadataConfig.newBuilder()
+        .enable(enableMetadataTable)
+        .withAssumeDatePartitioning(cfg.assumeDatePartitioning)
+        .build();
+
+    return FileSystemViewManager.createInMemoryFileSystemView(engineContext,
+        metaClient, metadataConfig);
+  }
+
+  public class AsyncMetadataTableValidateService extends HoodieAsyncService {
+    private final transient ExecutorService executor = Executors.newSingleThreadExecutor();
+
+    @Override
+    protected Pair<CompletableFuture, ExecutorService> startService() {
+      return Pair.of(CompletableFuture.supplyAsync(() -> {
+        while (true) {
+          try {
+            long start = System.currentTimeMillis();
+            doMetaTableValidation();
+            long toSleepMs = cfg.minValidateIntervalSeconds * 1000 - (System.currentTimeMillis() - start);
+
+            if (toSleepMs > 0) {
+              LOG.info("Last validate ran less than min validate interval: " + cfg.minValidateIntervalSeconds + " s, sleep: "
+                  + toSleepMs + " ms.");
+              Thread.sleep(toSleepMs);
+            }
+          } catch (HoodieValidationException e) {
+            LOG.error("Shutting down AsyncMetadataTableValidateService due to HoodieValidationException", e);
+            if (!cfg.ignoreFailed) {
+              throw e;
+            }
+          } catch (InterruptedException e) {
+            // ignore InterruptedException here.
+          }
+        }
+      }, executor), executor);
+    }
+  }
+
+  public static class FileSliceCompactor implements Comparator<FileSlice>, Serializable {
+
+    @Override
+    public int compare(FileSlice o1, FileSlice o2) {
+      return (o1.getPartitionPath() + o1.getFileId() + o1.getBaseInstantTime())
```

Comment:
Should there be any concatenator like '/' to mimic the full path?

Reply:
emmmm, maybe we don't need any concatenator. Just use  getPartitionPath, getFileId and getBaseInstantTime to flag the unique FileSlice and do sort based on it is fine.
Also we can use concatenator like '/' to mimic the full path but maybe there' s no diff for sort result :)

Reply:
Makes sense.  Let's keep it as is.